internal/backend/smb/smb.go

```diff
@@ -0,0 +1,557 @@
+	d, err := cn.smbShare.Open(dir)
+	if err != nil {
+		return err
+	}
+
+	sub, err := d.Readdirnames(-1)
+	if err != nil {
+		// ignore subsequent errors
+		_ = d.Close()
+		return err
+	}
+
+	err = d.Close()
+	if err != nil {
+		return err
+	}
+
+	for _, f := range sub {
+		err = b.visitFiles(ctx, cn, filepath.Join(dir, f), fn, true)
+		if err != nil {
+			return err
+		}
+	}
+	return ctx.Err()
+}
+
+func (b *Backend) visitFiles(ctx context.Context, cn *conn, dir string, fn func(restic.FileInfo) error, ignoreNotADirectory bool) error {
+	d, err := cn.smbShare.Open(dir)
+	if err != nil {
+		return err
+	}
+
+	if ignoreNotADirectory {
+		fi, err := d.Stat()
+		if err != nil || !fi.IsDir() {
+			// ignore subsequent errors
+			_ = d.Close()
+			return err
+		}
+	}
+
+	sub, err := d.Readdir(-1)
+	if err != nil {
+		// ignore subsequent errors
+		_ = d.Close()
+		return err
+	}
+
+	err = d.Close()
+	if err != nil {
+		return err
+	}
+
+	for _, fi := range sub {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		err := fn(restic.FileInfo{
+			Name: fi.Name(),
+			Size: fi.Size(),
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Delete removes the repository and all files.
+func (b *Backend) Delete(ctx context.Context) error {
+	debug.Log("Delete()")
+	cn, err := b.getConnection(ctx, b.ShareName)
+	if err != nil {
+		return err
+	}
+	defer b.putConnection(&cn)
+	return cn.smbShare.RemoveAll(b.Location())
+}
+
+// Close closes all open files.
+func (b *Backend) Close() error {
+	debug.Log("Close()")
+	err := b.drainPool()
+	return err
+}
+
+const (
+	pathSeparator = '/' // Always using '/' for SMB
+)
+
+// CreateTemp creates a new temporary file in the directory dir,
+// opens the file for reading and writing, and returns the resulting file.
+// The filename is generated by taking pattern and adding a random string to the end.
+// If pattern includes a "*", the random string replaces the last "*".
+// If dir is the empty string, CreateTemp uses the default directory for temporary files, as returned by TempDir.
+// Multiple programs or goroutines calling CreateTemp simultaneously will not choose the same file.
+// The caller can use the file's Name method to find the pathname of the file.
+// It is the caller's responsibility to remove the file when it is no longer needed.
+func (b *Backend) CreateTemp(cn *conn, dir, pattern string) (*smb2.File, error) {
+	if dir == "" {
+		dir = os.TempDir()
+	}
+
+	prefix, suffix, err := prefixAndSuffix(pattern)
+	if err != nil {
+		return nil, &fs.PathError{Op: "createtemp", Path: pattern, Err: err}
+	}
+	prefix = joinPath(dir, prefix)
+
+	try := 0
+	for {
+		name := prefix + nextRandom() + suffix
+		f, err := cn.smbShare.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0600)
+
+		if os.IsExist(err) {
+			if try++; try < 10000 {
+				continue
+			}
+			return nil, &fs.PathError{Op: "createtemp", Path: prefix + "*" + suffix, Err: fs.ErrExist}
+		}
+		return f, err
+	}
+}
```

Comment:
Just do what the SFTP backend does:

```
// tempSuffix generates a random string suffix that should be sufficiently long
// to avoid accidental conflicts.
func tempSuffix() string {
        var nonce [16]byte
        _, err := rand.Read(nonce[:])
        if err != nil {
                panic(err)
        }
        return hex.EncodeToString(nonce[:])
}

tmpFilename := filename + "-restic-temp-" + tempSuffix()
```

Reply:
Fixed